roqoqo/src/operations/analog_operations.rs

```diff
@@ -0,0 +1,116 @@
+// limitations under the License.
+
+use crate::operations::{
+    ImplementedIn1point10, InvolveQubits, InvolvedQubits, Operate, OperateSpinsAnalog, Substitute,
+    SupportedVersion,
+};
+use crate::RoqoqoError;
+use qoqo_calculator::CalculatorFloat;
+use std::collections::{HashMap, HashSet};
+use struqture::spins::{PauliProduct, SpinHamiltonian};
+use struqture::OperateOnDensityMatrix;
+use struqture::SpinIndex;
+
+/// Implements the continuous time, constant spin Hamiltonian
+#[derive(Debug, Clone, PartialEq, roqoqo_derive::Operate, roqoqo_derive::Substitute)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "json_schema", derive(schemars::JsonSchema))]
+pub struct ApplyConstantSpinHamiltonian {
+    /// Hamiltonian to be simulated.
+    pub hamiltonian: SpinHamiltonian,
+    /// Time of evolution.
+    pub time: CalculatorFloat,
+}
+
+#[allow(non_upper_case_globals)]
+const TAGS_ApplyConstantSpinHamiltonian: &[&str; 4] = &[
+    "Operation",
+    "ModeGateOperation",
+    "OperateSpinsAnalog",
+    "ApplyConstantSpinHamiltonian",
+];
+
+impl ImplementedIn1point10 for ApplyConstantSpinHamiltonian {}
+
+impl OperateSpinsAnalog for ApplyConstantSpinHamiltonian {
+    fn spin(&self) -> Vec<usize> {
+        let mut qubit_set = HashSet::new();
+        for pps in self.hamiltonian.keys() {
+            for (index, _) in pps.iter() {
+                qubit_set.insert(*index);
+            }
+        }
+        let mut qubits = Vec::from_iter(qubit_set);
+        qubits.sort();
+        qubits
+    }
+}
+impl SupportedVersion for ApplyConstantSpinHamiltonian {
+    fn minimum_supported_roqoqo_version(&self) -> (u32, u32, u32) {
+        (1, 10, 0)
+    }
+}
+
+impl InvolveQubits for ApplyConstantSpinHamiltonian {
+    fn involved_qubits(&self) -> InvolvedQubits {
+        InvolvedQubits::All
+    }
+}
+
+/// Implements the continuous time , time-dependent spin Hamiltonian
+#[derive(Debug, Clone, PartialEq, roqoqo_derive::Operate, roqoqo_derive::Substitute)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "json_schema", derive(schemars::JsonSchema))]
+pub struct ApplyTimeDependentSpinHamiltonian {
+    /// Hamiltonian to be simulated.
+    hamiltonian: SpinHamiltonian,
+    /// Time of evolution.
+    time: Vec<f64>,
+    values: HashMap<String, Vec<f64>>,
+}
+
+#[allow(non_upper_case_globals)]
+const TAGS_ApplyTimeDependentSpinHamiltonian: &[&str; 4] = &[
+    "Operation",
+    "ModeGateOperation",
+    "OperateSpinsAnalog",
+    "ApplyTimeDependentSpinHamiltonian",
+];
+
+impl ImplementedIn1point10 for ApplyTimeDependentSpinHamiltonian {}
+
+impl OperateSpinsAnalog for ApplyTimeDependentSpinHamiltonian {
+    fn spin(&self) -> Vec<usize> {
+        let mut qubit_set = HashSet::new();
+        for pps in self.hamiltonian.keys() {
+            for (index, _) in PauliProduct::iter(pps) {
```

Comment:
The fix used in the above implementation of this should be propagated (using `iter()`)